packages/iov-bns/src/bnsconnection.ts

```diff
@@ -467,61 +475,17 @@ export class BnsConnection implements BcpAtomicSwapConnection {
    * and then continuing with live feeds
    */
   public liveTx(txQuery: BcpTxQuery): Stream<ConfirmedTransaction> {
-    let updatesSubscription: Subscription | undefined;
+    const historyStream = fromListPromise(this.searchTx(txQuery));
+    const updatesStream = this.listenTx(txQuery);
+    const combinedStream = concat(historyStream, updatesStream);
 
-    const producer: Producer<ConfirmedTransaction> = {
-      start: listener => {
-        // called as soon as anyone subscribes to the stream
-
-        // 1. subscribe to updates before queying history to ensure we don't miss something; store updates in queue
-        // 2. when history is there, send history and process queue
-        // 3. skip queue and send updates directly
-
-        let doneSendingHistory = false;
-        const queue = new Array<ConfirmedTransaction>();
-
-        updatesSubscription = this.listenTx(txQuery).subscribe({
-          next: value => {
-            if (doneSendingHistory) {
-              listener.next(value);
-            } else {
-              queue.push(value);
-            }
-          },
-          error: error => listener.error(error),
-        });
-
-        this.searchTx(txQuery)
-          .then(history => {
-            for (const transaction of history) {
-              listener.next(transaction);
-            }
-
-            const historyIds = history.map(transaction => transaction.transactionId);
-
-            let element: ConfirmedTransaction | undefined;
-            // tslint:disable-next-line:no-conditional-assignment
-            while ((element = queue.shift())) {
-              const elementId = element.transactionId;
-              if (historyIds.indexOf(elementId) !== -1) {
-                // only do this for elements not already sent
-                listener.next(element);
-              }
-            }
+    // remove duplicates
+    const alreadySent = new Set<TransactionId>();
+    const deduplicatedStream = combinedStream
```

Comment:
I like this simplified logic quite a bit.
I don't quite understand what debug does, but it seems like it runs that code for each item and then passes the original item.

`debug(ct => alreadySent.add(ct.transactionId))` 
would be the same as
`filter(ct => {alreadySent.add(ct.transactionId); return true})`
but easier to read. Correct?

Reply:
So much code gone with this refactor. Like it!

Reply:
debug() simply runs a function for every event, without changing the stream. It was created for debugging purposes and has some simpler ways to call it:

* `.debug()` logs the event to console
* `.debug("label123")` logs the event to console, adding the debug label "label123"

For out use case `.debug(function_call)` would better be called `.forEarch(function_call)`.

My mid-term idea is to create a deduplication helper in @iov/stream and push the debug() call down the stack.